From f1a0e9a75bf2af49659a25bf2c729af20d1cb24d Mon Sep 17 00:00:00 2001
From: Ivan Shvedunov <ivan.shvedunov@travelping.com>
Date: Thu, 4 Feb 2021 21:21:30 +0300
Subject: [PATCH] ping: fix aborting on keypress

Type: fix

Currently ping stops on events like SOCKET_READ_EVENT,
which makes it hard to use over e.g. govpp as it aborts
immediately most of the time. With this patch, ping only
stops upon real CLI read / quit events.

Signed-off-by: Ivan Shvedunov <ivan4th@gmail.com>
Change-Id: Id7a8d0b0fdeb7bbc7b85240e398d27bd5199345b
---
 src/plugins/ping/ping.c | 5 +++--
 src/vlib/unix/cli.c     | 7 -------
 src/vlib/unix/unix.h    | 8 ++++++++
 3 files changed, 11 insertions(+), 9 deletions(-)

diff --git a/src/plugins/ping/ping.c b/src/plugins/ping/ping.c
index f56f44ffb..a9136cc11 100644
--- a/src/plugins/ping/ping.c
+++ b/src/plugins/ping/ping.c
@@ -16,6 +16,7 @@
 #include <stddef.h>
 
 #include <vlib/vlib.h>
+#include <vlib/unix/unix.h>
 #include <vnet/fib/ip6_fib.h>
 #include <vnet/fib/ip4_fib.h>
 #include <vnet/fib/fib_entry.h>
@@ -1175,11 +1176,11 @@ run_ping_ip46_address (vlib_main_t * vm, u32 table_id, ip4_address_t * pa4,
 		  }
 	      }
 	      break;
-	    default:
+	    case UNIX_CLI_PROCESS_EVENT_READ_READY:
+	    case UNIX_CLI_PROCESS_EVENT_QUIT:
 	      /* someone pressed a key, abort */
 	      vlib_cli_output (vm, "Aborted due to a keypress.");
 	      goto double_break;
-	      break;
 	    }
 	  vec_free (event_data);
 	}
diff --git a/src/vlib/unix/cli.c b/src/vlib/unix/cli.c
index 5979dbdbd..470744fb8 100644
--- a/src/vlib/unix/cli.c
+++ b/src/vlib/unix/cli.c
@@ -449,13 +449,6 @@ static unix_cli_parse_actions_t unix_cli_parse_pager[] = {
 
 #undef _
 
-/** CLI session events. */
-typedef enum
-{
-  UNIX_CLI_PROCESS_EVENT_READ_READY,  /**< A file descriptor has data to be read. */
-  UNIX_CLI_PROCESS_EVENT_QUIT,	      /**< A CLI session wants to close. */
-} unix_cli_process_event_type_t;
-
 /** CLI session telnet negotiation timer events. */
 typedef enum
 {
diff --git a/src/vlib/unix/unix.h b/src/vlib/unix/unix.h
index 9fa95a0c1..86ef88b61 100644
--- a/src/vlib/unix/unix.h
+++ b/src/vlib/unix/unix.h
@@ -109,6 +109,14 @@ typedef struct
 
 } unix_main_t;
 
+/** CLI session events. */
+typedef enum
+{
+  UNIX_CLI_PROCESS_EVENT_READ_READY, /**< A file descriptor has data to be
+					read. */
+  UNIX_CLI_PROCESS_EVENT_QUIT,	     /**< A CLI session wants to close. */
+} unix_cli_process_event_type_t;
+
 /* Global main structure. */
 extern unix_main_t unix_main;
 extern clib_file_main_t file_main;
-- 
2.28.0

